docs/features/incremental-generators.md

```diff
@@ -0,0 +1,478 @@
+# Incremental Generators
+
+## Summary
+
+Incremental generators are intended to be a new API that exists alongside
+[source generators](generators.md) to allow users to specify generation
+strategies that can be applied in a high performance way by the hosting layer.
+
+### High Level Design Goals
+
+- Allow for a finer grained approach to defining a generator
+- Scale source generators to support 'Roslyn/CoreCLR' scale projects in Visual Studio
+- Exploit caching between fine grained steps to reduce duplicate work
+- Support generating more items that just source texts
+- Exist alongside `ISourceGenerator` based implementations
+
+## Implementation
+
+An incremental generator is an implementation of `Microsoft.CodeAnalysis.IIncrementalGenerator`.
+
+```csharp
+namespace Microsoft.CodeAnalysis
+{
+    public interface IIncrementalGenerator
+    {
+        void Initialize(IncrementalGeneratorInitializationContext initContext);
+    }
+}
+```
+
+As with source generators, incremental generators are defined in external
+assemblies and passed to the compiler via the `-analyzer:` option.
+Implementations are required to be annotated with the
+`Microsoft.CodeAnalysis.GeneratorAttribute` with an optional parameter
+indicating the languages the generator supports:
+
+```csharp
+[Generator(LanguageNames.CSharp)]
+public class MyGenerator : IIncrementalGenerator { ... }
+```
+
+An assembly can contain a mix of diagnostic analyzers, source generators and
+incremental generators.
+
+### Initialization
+
+`IIncrementalGenerator` has an `Initialize` method that is called by the host
+(either the IDE or the command-line compiler) exactly once, regardless of the
+number of further compilations that may occur. `Initialize` passes an instance
+of `IncrementalGeneratorInitializationContext` which can be used by the
+generator to register a set of callbacks that affect how future generation
+passes will occur.
+
+Currently `IncrementalGeneratorInitializationContext` supports two callbacks:
+
+- `RegisterForPostInitialization(GeneratorPostInitializationContext)`: the same
+  callback in shape and function as supported by Source Generators today
+- `RegisterExecutionPipeline(IncrementalGeneratorPipelineContext)`: replaces Execute, described below
+
+### Pipeline based execution
+
+Rather than a dedicated `Execute` method, an Incremental Generator instead
+creates an execution 'pipeline' as part of the initialization process via the
+`RegisterExecutionPipeline` method:
+
+```csharp
+public void Initialize(IncrementalGeneratorInitializationContext initContext)
+{
+    initContext.RegisterExecutionPipeline(context =>
+    {
+        // build the pipeline...
+    });
+}
+```
+
+This pipeline is not directly executed, but instead consists of a set of steps
```

Comment:
>a set of steps [](http://example.com/codeflow?start=64&length=14)

nit: consider something more specific like "a tree of steps, or "a DAG composed of steps". Also it's worth mentioning that the graph is static (doesn't depend or change based on execution of the transformation steps).